Reduce document id size from 64bits to 32bits

Kerollmops opened the pull request:
This is another step towards improving the performances and disk space usage of MeiliSearch.

Kerollmops commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+        }
+        let user_ids = build.into_inner().unwrap();
+
+        // TODO prefer using self.put_user_ids
+        self.main.put::<_, Str, ByteSlice>(writer, USER_IDS_KEY, user_ids.as_slice())
```
I kept that because we are not using the version 0.4.0 of the fst library and the function [`into_map` that I added to the `MapBuilder`](https://docs.rs/fst/0.4.3/fst/struct.MapBuilder.html#method.into_map) are, therefore, not available.
It is easier to directly write the bytes instead of constructing an `fst::Map` and passing it to the `put_user_ids`.

Moving to fst 0.4.0 is already planned in the tracking issue #692.

Kerollmops commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+        self.put_internal_ids(writer, &internal_ids)
+    }
+
+    pub fn put_user_ids(self, writer: &mut heed::RwTxn<MainT>, ids: &fst::Map) -> ZResult<()> {
```
User ids are the documents ids given by the user, those are always converted to strings and put inside of this `fst::Map`, as you already guessed.

qdequele replied:
`users_ids` remind me of an Identifier to track the user during all these requests. What do you think about `internal_doc_id` and `external_doc_id`?

MarinPostma replied:
I agree with @qdequele, We need something more specific, like `(put_)internal_docids` vs. `(put_)user_docids`

MarinPostma replied:
It should be clear that it's docids we're dealing with

Kerollmops replied:
Yeah, I am ok with `external` and `internal` prefixes, I change that! ;)

Kerollmops replied:
The code is much more clear now, thank you both.
Even better: "external" and "internal" are the same length, I can align function calls and variable declarations!

Kerollmops commented:
Yeah, I understand that consfusion, I will introduce the function you propose, that is easier to use!
I was using this `DiscoverIds` Iterator because this code is a port from the daugt codebase and daugt only support CSVs with unknown length, therefore the function you propose would have been useless.

Kerollmops commented:
In fact, no, I will not add this `fn discover_internal_ids(len: usize) -> Vec<Id>` method that we talk about, this `DiscoverIds` struct is easy enough to use and help us keep good performances by not searching for too much ids that will not be used (when you replace many documents which already have ids).
I keep it like that, the `DiscoverIds` Iterator [is used in only one place][1], so... Not that hard to understad how its used.

[1]: https://github.com/meilisearch/MeiliSearch/pull/696/files#diff-cfebcd6ba57d3f5acc05437fb7f99b8fR164

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -150,17 +150,26 @@ pub fn apply_addition<'a, 'b>(
 ) -> MResult<()> {
     let mut documents_additions = HashMap::new();
+    let mut new_user_ids = BTreeMap::new();
+    let mut new_internal_ids = Vec::with_capacity(new_documents.len());
```
```suggestion
    let mut new_user_docids = BTreeMap::new();
    let mut new_internal_docids = Vec::with_capacity(new_documents.len());
```

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -150,17 +150,26 @@ pub fn apply_addition<'a, 'b>(
-        let document_id = extract_document_id(&primary_key, &document)?;
+        let (document_id, userid) = extract_document_id(&primary_key, &document, &user_ids, &mut available_ids)?;
+        new_user_ids.insert(userid, document_id.0);
+        new_internal_ids.push(document_id);
```
we don't really follow the declaration on top of scope rule, so I think it's more readable if declarations are as close a possible from first use, wdyt?

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -181,9 +190,9 @@ pub fn apply_addition<'a, 'b>(
+    // 2. remove the documents postings lists
     let number_of_inserted_documents = documents_additions.len();
-    let documents_ids = documents_additions.iter().map(|(id, _)| *id).collect();
+    let documents_ids = new_user_ids.iter().map(|(userid, _)| userid.clone()).collect();
```
```suggestion
    let documents_ids = new_user_docids.iter().map(|(docid, _)| docid.clone()).collect();
```

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -150,17 +150,26 @@ pub fn apply_addition<'a, 'b>(
     for mut document in new_documents {
-        let document_id = extract_document_id(&primary_key, &document)?;
+        let (document_id, userid) = extract_document_id(&primary_key, &document, &user_ids, &mut available_ids)?;
+        new_user_ids.insert(userid, document_id.0);
```
is `document_id` internal? I'm lost now :(

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -233,6 +242,11 @@ pub fn apply_addition<'a, 'b>(
+    let new_user_ids = fst::Map::from_iter(new_user_ids.iter().map(|(u, i)| (u, *i as u64)))?;
+    let new_internal_ids = sdset::SetBuf::from_dirty(new_internal_ids);
+    index.main.merge_user_ids(writer, &new_user_ids)?;
+    index.main.merge_internal_ids(writer, &new_internal_ids)?;
```
same with docids. Also, `u` and `i` are not very helpful telling what this is, could you use more explicit names?

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -31,7 +31,7 @@ impl DocumentsDeletion {
     }
 
-    pub fn delete_document_by_id(&mut self, document_id: DocumentId) {
+    pub fn delete_document_by_user_id(&mut self, document_id: String) {
```
```suggestion
    pub fn delete_document_by_user_docid(&mut self, document_id: String) {
```

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -57,7 +57,7 @@ pub fn push_documents_deletion(
     updates_store: store::Updates,
     updates_results_store: store::UpdatesResults,
-    deletion: Vec<DocumentId>,
+    deletion: Vec<String>,
```
```suggestion
    user_doicds: Vec<String>,
```

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -70,8 +70,24 @@ pub fn push_documents_deletion(
     index: &store::Index,
-    deletion: Vec<DocumentId>,
-) -> MResult<()> {
+    deletion: Vec<String>,
```
```suggestion
    user_docids: Vec<String>,
```

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -70,8 +70,24 @@ pub fn push_documents_deletion(
+
+        let new_user_ids = fst::Map::from_iter(new_user_ids.into_iter().map(|k| (k, 0))).unwrap();
+        (new_user_ids, SetBuf::from_dirty(internal_ids))
+    };
```
k is docid right?
```suggestion
    let (user_docids, internal_docids) = {
        let new_user_docids = SetBuf::from_dirty(deletion);
        let mut internal_docids = Vec::new();

        let user_ids = index.main.user_docids(writer)?;
        for user_docid in new_user_docids.as_slice() {
            if let Some(id) = user_ids.get(userid) {
                internal_docids.push(DocumentId(id as u32));
            }
        }

        let new_user_docids = fst::Map::from_iter(new_user_docids.into_iter().map(|docid| (docid, 0))).unwrap();
        (new_user_docids, SetBuf::from_dirty(internal_docids))
    };
```

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -84,16 +100,15 @@ pub fn apply_documents_deletion(
     // facet filters deletion
     if let Some(attributes_for_facetting) = index.main.attributes_for_faceting(writer)? {
-        let facet_map = facets::facet_map_from_docids(writer, &index, &deletion, &attributes_for_facetting)?;
+        let facet_map = facets::facet_map_from_docids(writer, &index, &internal_ids, &attributes_for_facetting)?;
```
```suggestion
        let facet_map = facets::facet_map_from_docids(writer, &index, &internal_docids, &attributes_for_facetting)?;
```

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -84,16 +100,15 @@ pub fn apply_documents_deletion(
-    let idset = SetBuf::from_dirty(deletion);
     let mut words_document_ids = HashMap::new();
-    for id in idset {
+    for id in internal_ids.iter().cloned() {
```
```suggestion
    for id in internal_docids.iter().cloned() {
```

MarinPostma commented on meilisearch-core/src/update/helpers.rs:
```diff
@@ -96,28 +95,43 @@ pub fn value_to_number(value: &Value) -> Option<Number> {
+/// Validates a string representation to be a correct document id and returns
+/// the corresponding id or generate a new one, this is the way we produce documents ids.
+pub fn discover_document_id(
+    userid: &str,
+    user_ids: &fst::Map,
+    available_ids: &mut DiscoverIds<'_>,
```
```suggestion
    user_docid: &str,
    user_docids: &fst::Map,
    available_docids: &mut DiscoverIds<'_>,
```

MarinPostma commented on meilisearch-http/src/routes/document.rs:
```diff
@@ -43,11 +43,16 @@ async fn get_document(
+    let internal_id = match internal_id {
+        Some(internal_id) => internal_id,
+        None => return Err(ResponseError::document_not_found(&path.document_id)),
+    };
```
```suggestion
    let internal_docid = index.main.user_to_internal_docid(&reader, &path.document_id)?
    .ok_or(ResponseError::document_not_found(&path.document_id))?;
```

Kerollmops commented:
@MarinPostma Yeah, sorry I already updated all of those lines and changed the `user_ids` into `external_docids` it is clear and fits really good! Thank you, but you didn't had to make it yourself :)

MarinPostma commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+        let external_docids = self.external_docids(writer)?;
+        let mut op = external_docids.op().add(new_ids.into_stream()).r#union();
+        let mut build = fst::MapBuilder::memory();
+        while let Some((userid, values)) = op.next() {
+            build.insert(userid, values[0].value).unwrap();
```
```suggestion
        while let Some((docid, values)) = op.next() {
            build.insert(docid, values[0].value).unwrap();
```

MarinPostma commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+        let external_docids = self.external_docids(writer)?;
+        let mut op = external_docids.op().add(ids.into_stream()).difference();
+        let mut build = fst::MapBuilder::memory();
+        while let Some((userid, values)) = op.next() {
+            build.insert(userid, values[0].value).unwrap();
```
```suggestion
        while let Some((docid, values)) = op.next() {
            build.insert(docid, values[0].value).unwrap();
```

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -233,6 +243,11 @@ pub fn apply_addition<'a, 'b>(
 
     index.main.put_schema(writer, &schema)?;
 
+    let new_external_docids = fst::Map::from_iter(new_external_docids.iter().map(|(u, i)| (u, *i as u64)))?;
```
it is not clear what `u` and `i` are, even more now that you changed userid to external_docid

MarinPostma commented on meilisearch-core/src/update/documents_deletion.rs:
```diff
@@ -70,8 +70,24 @@ pub fn push_documents_deletion(
+        let new_external_docids = SetBuf::from_dirty(external_docids);
+        let mut internal_docids = Vec::new();
+
+        let user_ids = index.main.external_docids(writer)?;
+        for userid in new_external_docids.as_slice() {
+            if let Some(id) = user_ids.get(userid) {
+                internal_docids.push(DocumentId(id as u32));
+            }
+        }
```
```suggestion
        let old_external_docids = index.main.external_docids(writer)?;
        for external_docid in new_external_docids.as_slice() {
            if let Some(id) = old_external_docids.get(external_docid) {
                internal_docids.push(DocumentId(id as u32));
            }
        }
```

MarinPostma commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+        self.main.put::<_, Str, ByteSlice>(writer, EXTERNAL_DOCIDS_KEY, ids.as_fst().as_bytes())
+    }
+
+    pub fn merge_external_docids(self, writer: &mut heed::RwTxn<MainT>, new_ids: &fst::Map) -> ZResult<()> {
```
```suggestion
    pub fn merge_external_docids(self, writer: &mut heed::RwTxn<MainT>, new_external_docids: &fst::Map) -> ZResult<()> {
```

MarinPostma replied:
the underlying type is hidden by the fst::map

MarinPostma commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+    pub fn merge_external_docids(self, writer: &mut heed::RwTxn<MainT>, new_ids: &fst::Map) -> ZResult<()> {
+        use fst::{Streamer, IntoStreamer};
+
+        // Do an union of the old and the new set of user ids.
```
```suggestion
        // Do an union of the old and the new set of external docids.
```

MarinPostma commented on meilisearch-core/src/store/main.rs:
```diff
@@ -71,9 +74,90 @@ impl Main {
+
+        // Do an union of the old and the new set of user ids.
+        let external_docids = self.external_docids(writer)?;
+        let mut op = external_docids.op().add(new_ids.into_stream()).r#union();
```
```suggestion
        let mut op = external_docids.op().add(new_external_docids.into_stream()).r#union();
```

MarinPostma commented on meilisearch-core/src/update/documents_addition.rs:
```diff
@@ -148,19 +148,29 @@ pub fn apply_addition<'a, 'b>(
+
     for mut document in new_documents {
-        let document_id = extract_document_id(&primary_key, &document)?;
+        let (document_id, userid) = extract_document_id(&primary_key, &document, &external_docids, &mut available_ids)?;
+        new_external_docids.insert(userid, document_id.0);
```
I'm confused with this one, what's document_id and what's userid
